### PR TITLE
Expose vds interpolation in api

### LINKED ISSUE #14: Interpolation
OpenVDS offers several interpolation methods. Expose (all?) of these as an optional parameter and decide on and document a default.

### PR DESCRIPTION
**Pull request by LeikvollE**
We know that that vds interpolation has some known bugs, but we still
expose it as is.

### CODE REVIEW

File: internal/vds/vds.cpp [hunk 1]
```diff
@@ -419,7 +433,8 @@ struct vdsbuffer fetch_fence(
     const std::string& credentials,
     const std::string& coordinate_system,
     const float* coordinates,
-    size_t npoints
+    size_t npoints,
+    enum interpolationMethod interpolationMethod
```
Comment (achaikou) [inline internal/vds/vds.cpp:437]: This is already a cpp file and chosen convention is a default snake case (coordinate_system). So interpolation_method is better.

**Review by achaikou — COMMENTED** [review]
Great, I checked against unmerged tests, everything works nicely!

With regards to PR description: I think it closes just part of the issue.
The full issue is to figure out what different interpolation options are and which one we better use and wtf is going on in there in general. (and also document stuff for the user, which is also not done)

So maybe just reference this PR in the issue, but do not close it?

File: internal/vds/vds.h [hunk 2]
```diff
@@ -23,6 +23,14 @@ enum axis {
     SAMPLE    = 7,
 };
 
+enum interpolationMethod {
```
[the remark this answers is not in the document]
Comment (LeikvollE) [reply internal/vds/vds.h]: After a short discussion with @ErlendHaa, it looks like we are not able to expose openvds directly to go because it is c++.

File: internal/vds/vds_test.go [hunk 3]
```diff
@@ -408,16 +411,97 @@ func TestFence(t *testing.T) {
...
+			expected: []float32{
+				100, 101, 102, 103,
+				106, 107, 108, 109,
+				116, 117, 118, 119,
+				0, 0, 0, 0,
+			},
+			method: "linear",
+		},
+		{
+			expected: []float32{
```
[the remark this answers is not in the document]
Comment (LeikvollE) [reply internal/vds/vds_test.go]: The expected numbers were generated from a separate script calling open vds directly. The idea was the the test was supposed to test our handling of the interpolation and not the whether open vds does anything sensible with the data we pass to it.

I implemented the four other tests you suggested.

File: internal/vds/vds.go [hunk 4]
```diff
@@ -42,6 +42,27 @@ func GetAxis(direction string) (int, error) {
...
+	case "angular":
+		return C.ANGULAR, nil
+	case "triangular":
+		return C.TRIANGULAR, nil
+	case "":
+		return C.NEAREST, nil
+	default:
+		options := "nearest, linear, cubic, angular or triangular"
+		msg := "Invalid interpolation method '%s', valid options are: %s"
+		return 0, errors.New(fmt.Sprintf(msg, interpolation, options))
```
[the remark this answers is not in the document]
Comment (LeikvollE) [reply internal/vds/vds.go]: Changed to -1. The reason for the 0 value was that when the error is any value other than nil the first value would not be valid anyways.

File: internal/vds/vds.go [hunk 5]
```diff
@@ -42,6 +42,27 @@ func GetAxis(direction string) (int, error) {
...
+	case "linear":
+		return C.LINEAR, nil
+	case "cubic":
+		return C.CUBIC, nil
+	case "angular":
+		return C.ANGULAR, nil
+	case "triangular":
+		return C.TRIANGULAR, nil
+	case "":
+		return C.NEAREST, nil
```
[the remark this answers is not in the document]
Comment (LeikvollE) [reply internal/vds/vds.go]: Fixed

File: internal/vds/vds_test.go [hunk 6]
```diff
@@ -427,81 +427,55 @@ func TestInvalidFence(t *testing.T) {
...
+	if defaultInterpolation != nearestInterpolation {
+		msg := "[fence_interpolation]Default interpolation is not nearest"
+		t.Error(msg)
+	}
+}
 
+func TestInvalidInterpolationMethod(t *testing.T) {
+	interpolation := "sand"
+	_, err := GetInterpolationMethod(interpolation)
+	if err == nil {
```
Comment (achaikou) [inline internal/vds/vds_test.go:468]: Small comment, feel free to disregard if you don't like it, but I would vote for additionally testing important part of error message, like:
```
if !strings.Contains(err.Error(), "Invalid interpolation method")
  complain message is wrong
```

Whether to test or not error messages is often a personal preference. I by default prefer to do so, but many do not.

**Review by achaikou — COMMENTED** [review]
I hope you have been informed that I am evil :smile:

File: internal/vds/vds_test.go [hunk 7]
```diff
@@ -427,81 +427,55 @@ func TestInvalidFence(t *testing.T) {
...
+	interpolationMethods := []string{"nearest", "linear", "cubic", "triangular", "angular"}
+	for i, v1 := range interpolationMethods {
+		interpolationMethod, _ := GetInterpolationMethod(v1)
+		buf1, _ := Fence(well_known, "cdp", fence, interpolationMethod)
+		for j, v2 := range interpolationMethods {
+			if i <= j {
+				continue
+			}
+			interpolationMethod, _ := GetInterpolationMethod(v2)
+			buf2, _ := Fence(well_known, "ij", fence, interpolationMethod)
```
[the remark this answers is not in the document]
Comment (LeikvollE) [reply internal/vds/vds_test.go:441]: Small oversight when changing from ij to cdp coordinates.

File: internal/vds/vds_test.go (hunk 6 above)
Comment (LeikvollE) [reply internal/vds/vds_test.go:472]: I added this as i can see it has been done in other tests

File: internal/vds/vds_test.go [hunk 8]
```diff
@@ -408,16 +411,82 @@ func TestFence(t *testing.T) {
...
+
+func TestInvalidInterpolationMethod(t *testing.T) {
+	interpolation := "sand"
+	_, err := GetInterpolationMethod(interpolation)
+	if err == nil {
+		msg := "[fence_interpolation]Expected fail given invalid interpolation method: %v"
+		t.Errorf(msg, interpolation)
+	} else {
+		options := "nearest, linear, cubic, angular or triangular"
+		expected := fmt.Sprintf("Invalid interpolation method 'sand', valid options are: %s", options)
```
Comment (achaikou) [inline internal/vds/vds_test.go:477]: Because I like to point out random useless stuff :smile: 
You don't actually need to import fmt module (as you mentioned, there are other tests in this file which check for errors - and none of them used fmt)

But this is absolutely valid approach, so leave it :)

**Review by achaikou — COMMENTED** [review]
I can't find anything more what to complain about, so prepare for merge!!
1. Rebase and squash fixups
2. Consider again if this PR closes the interpolation issue or do we want to keep it open for when we figure openvds interpolation out (or if we want to create a separate new issue).

File: internal/vds/vds_test.go (hunk 8 above)
Comment (LeikvollE) [reply internal/vds/vds_test.go:464]: Rebased and squashed

**Review by achaikou — APPROVED** [review]
